Add arm64 support for EnC

jakobbotsch opened the pull request:
This adds support for EnC on arm64. A couple of notes on the
implementation compared to x64:
- On x64 we get the fixed stack size from unwind info. However, for the
  frames we set up on arm64 for EnC it is not possible to extract the
  frame size from there because their prologs generally look like

  stp fp, lr, [sp,#-16]!
  mov fp, sp
  sub sp, sp, # 144

  with unwind codes like the following:

  set_fp; mov fp, sp

  save_fplr_x # 1 (0x01); stp fp, lr, [sp, #-16]!

  As can be seen, it is not possible to get the fixed stack size from
  unwind info in this case. Instead we pass it through the GC info that
  already has a section for EnC data.

- On arm64 the JIT is required to place the PSPSym at the same offset
  from caller-SP for both the main function and for funclets. Due to
  this we try to allocate the PSPSym as early as possible in the main
  function and we must take some care in funclets.  However, this
  conflicts with the EnC frame header that the JIT uses to place values
  that must be preserved on EnC transitions. This is currently
  callee-saved registers and the MonitorAcquired boolean.

  Before this change we were allocating PSPSym above (before) the
  monitor acquired boolean, but we now have to allocate MonitorAcquired
  first, particularly because the size of the preserved header cannot
  change on EnC transitions, while the PSPSym can disappear or appear.
  This changes frame allocation slightly for synchronized functions.

jakobbotsch committed:
Add arm64 support for EnC

This adds support for EnC on arm64. A couple of notes on the
implementation compared to x64:
- On x64 we get the fixed stack size from unwind info. However, for the
  frames we set up on arm64 for EnC it is not possible to extract the
  frame size from there because their prologs generally look like

  stp fp, lr, [sp,#-16]!
  mov fp, sp
  sub sp, sp, #144

  with unwind codes like the following:

  set_fp; mov fp, sp

  save_fplr_x #1 (0x01); tp fp, lr, [sp, #-16]!

  As can be seen, it is not possible to get the fixed stack size from
  unwind info in this case. Instead we pass it through the GC info that
  already has a section for EnC data.

- On arm64 the JIT is required to place the PSPSym at the same offset
  from caller-SP for both the main function and for funclets. Due to
  this we try to allocate the PSPSym as early as possible in the main
  function and we must take some care in funclets.  However, this
  conflicts with the EnC frame header that the JIT uses to place values
  that must be preserved on EnC transitions. This is currently
  callee-saved registers and the MonitorAcquired boolean.

  Before this change we were allocating PSPSym above (before) the
  monitor acquired boolean, but we now have to allocate MonitorAcquired
  first, particularly because the size of the preserved header cannot
  change on EnC transitions, while the PSPSym can disappear or appear.
  This changes frame allocation slightly for synchronized functions.

jakobbotsch committed:
Run jit-format

AndyAyersMS commented on src/coreclr/vm/encee.cpp:
```diff
@@ -694,7 +713,7 @@ HRESULT EditAndContinueModule::ResumeInUpdatedFunction(
     EEPOLICY_HANDLE_FATAL_ERROR(CORDBG_E_ENC_INTERNAL_ERROR);
     return hr;
-#endif // #if define(TARGET_ARM64) || defined(TARGET_ARM)
+#endif // #if defined(TARGET_ARM)
```
```suggestion
#endif // #if defined(TARGET_ARM) || defined(TARGET_LOONGARCH64)
```

AndyAyersMS reviewed (approved):
Do we need to update the R2R version because of the gc info change? I suppose we don't as this new format is only there with ENC, which we won't see in R2R.

jakobbotsch commented:
> Do we need to update the R2R version because of the gc info change? I suppose we don't as this new format is only there with ENC, which we won't see in R2R.

Right, I checked with @davidwrighton and he was under the same belief.

jakobbotsch commented:
Looks like there is an OSR + synchronized assert I'll need to sort out.

jakobbotsch committed:
Feedback

jakobbotsch committed:
Fix funclets for OSR methods with monitor-acquired bool

OSR methods reuse the bool created by the tier 0 frame so we should not
allocate space for it.

tommcdon reviewed (approved):
LGTM, thanks!
